File: Data Structures & Algorithms/house-robber/submission-0.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        mp = {}
        
        for i in range(len(nums) - 1, -1, -1):
            if i == len(nums) - 1:
                mp[i] = nums[i]
                continue
            
            mp[i] = max(nums[i] + mp.get(i + 2, 0), mp[i + 1])

        return mp[0]
            
            

File: Data Structures & Algorithms/maximum-product-subarray/submission-0.java
class Solution {
    public int maxProduct(int[] nums) {
        int n = nums.length;
        int dp[][] = new int[n][n];
        int largest = Integer.MIN_VALUE;
        for (int len = 0; len < n; len++) {
            for (int i = 0; i < n; i++) {
                int j = i + len;
                if (j >= n) continue;
                if (i == j) {
                    dp[i][j] = nums[i];
                } else {
                    dp[i][j] = dp[i][j - 1] * nums[j];
                }
                largest = Math.max(largest, dp[i][j]);
            }
        }
        return largest;

    }
}
